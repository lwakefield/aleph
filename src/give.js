import { h, Component } from 'preact'

import Stores from './stores'

const pairs = obj => Object.keys(obj).map(k => [k, obj[k]])

export default function give (propMapFn) {
	return target => connect(target, propMapFn)
}

export function connect (Target, propMapFn) {

	return class Connector extends Component {
		constructor (...args) {
			super(...args)

			this.__listeners = new Map()
			this.propMap = propMapFn(this.props)

			for (const [key, path] of pairs(this.propMap)) {
				this.__connect(key, path)
			}
		}

		componentDidUpdate () {
			const nextPropMap = propMapFn(this.props)
			const lastPropMap = this.propMap

			for (const [key, path] of pairs(nextPropMap)) {
				if (lastPropMap[key] !== path) {
					this.__disconnect(lastPropMap[key])
					this.__connect(key,  path)
				}
			}

			this.propMap = nextPropMap
		}

		__connect (key, path) {
			const fn = () => this.setState({[key]: Stores.get(path)})
			fn()
			Stores.on(path, fn)
			this.__listeners.set(path, fn)
		}

		__disconnect (path) {
			const listener = this.__listeners.get(path)
			Stores.off(path, listener)
			this.__listeners.delete(path)
		}

		render (props, state) {
			return <Target {...props} {...state} />
		}
	}

}
