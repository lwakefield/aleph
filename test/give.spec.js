import {h, Component, render } from 'preact'
import {spy, mock, stub} from 'sinon'

import give from 'give'

afterEach(() => give.__ResetDependency__('Stores'))

test('give() creates a Connector', () => {
    @give()
    class Foo extends Component { }

    expect(Foo.prototype.constructor.name).toEqual('Connector')
    expect(Foo.prototype.__connect).toBeDefined()
    expect(Foo.prototype.__disconnect).toBeDefined()
})

test('Connector instantiates correctly', () => {
    const fn = stub().returns({foo: 'bar', baz: 'qux'})

    @give(fn)
    class Foo extends Component { }
    spy(Foo.prototype, '__connect')

    const inst = new Foo()

    expect(inst).toMatchSnapshot()
    expect(fn.callCount).toEqual(1)
    expect(inst.__connect.callCount).toEqual(2)
    expect(inst.__connect.calledWith('foo', 'bar')).toBeTruthy()
    expect(inst.__connect.calledWith('baz', 'qux')).toBeTruthy()
})

test('Connector.__connect works correctly', () => {
    @give(() => ({}))
    class Foo extends Component { }

    const inst = new Foo()
    spy(inst, 'setState')
    const Stores = {
        on: stub(),
        get: stub().returns('hello world')
    }
    give.__Rewire__('Stores', Stores)

    inst.__connect('foo', 'bar')
    expect(inst.setState.calledOnce).toBeTruthy()
    expect(inst.setState.calledWith({foo: 'hello world'})).toBeTruthy()
    expect(Stores.on.calledOnce)
    expect(Stores.on.getCall(0).args[0]).toEqual('bar')
    expect(Stores.on.getCall(0).args[1]).toBeInstanceOf(Function)
    expect(inst.__listeners.get('bar')).toBeInstanceOf(Function)
})

test('Connector.__disconnect works correctly', () => {
    @give(() => ({}))
    class Foo extends Component { }

    const inst = new Foo()
    const fn = stub()
    inst.__listeners.set('foo', fn)
    const Stores = {
        off: stub(),
        get: stub().returns('hello world')
    }
    give.__Rewire__('Stores', Stores)

    inst.__disconnect('foo')
    expect(Stores.off.calledOnce)
    expect(Stores.off.calledWith('foo', fn)).toBeTruthy()
    expect(inst.__listeners.get('foo')).toBeUndefined()
})

test('Connector.componentDidUpdate works correctly', () => {
    // will pass through props
    @give(v => v)
    class Foo extends Component { }

    const inst = new Foo({
        foo: 'bar',
        baz: 'qux',
    })
    inst.props = {
        foo: 'bar1',
        baz: 'qux'
    }
    spy(inst, '__disconnect')
    spy(inst, '__connect')

    // should update foo, but leave baz alone
    inst.componentDidUpdate()
    expect(inst.propMap).toEqual({foo: 'bar1', baz: 'qux'})
    expect(inst.__disconnect.calledOnce).toBeTruthy()
    expect(inst.__disconnect.calledWith('bar')).toBeTruthy()
    expect(inst.__connect.calledOnce).toBeTruthy()
    expect(inst.__connect.calledWith('foo', 'bar1')).toBeTruthy()
})

test('Connector.render works correctly', () => {
    @give(() => ({}))
    class Foo extends Component { }

    const inst = new Foo()
    const rendered = inst.render({foo: 'bar'}, {baz: 'qux'})
    expect(rendered).toMatchSnapshot()
})

test('Connector.render renders to document', () => {
    @give(() => ({
        foo: 'this is my foo',
        bar: 'this is my bar',
    }))
    class Foo extends Component {
        render ({foo, bar}) {
            return <p>my foo: {foo}, mybar: {bar}</p>
        }
    }

    const Stores = {
        on: stub(),
        off: stub(),
        get: v => v
    }
    give.__Rewire__('Stores', Stores)

    render(<Foo />, document.body)
    expect(document.body.innerHTML).toMatchSnapshot()
})

test('Connector receives updates from Stores', () => {
    const data = {bar: 'hello from bar'}
    const listeners = {}
    const Stores = {
        on: (k, v) => listeners[k] = v,
        get: k => data[k]
    }
    give.__Rewire__('Stores', Stores)

    @give(() => ({'foo': 'bar'}))
    class Foo extends Component { }

    const inst = new Foo()
    expect(inst.state).toEqual({foo: 'hello from bar'})
    data.bar = 'hello again from bar'
    listeners.bar()
    expect(inst.state).toEqual({foo: 'hello again from bar'})
})
