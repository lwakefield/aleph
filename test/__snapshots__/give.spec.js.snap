// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connector instantiates correctly 1`] = `
Connector {
  "__d": true,
  "__listeners": Map {
    "bar" => [Function],
    "qux" => [Function],
  },
  "__s": Object {},
  "context": undefined,
  "propMap": Object {
    "baz": "qux",
    "foo": "bar",
  },
  "props": undefined,
  "state": Object {
    "baz": undefined,
    "foo": undefined,
  },
}
`;

exports[`Connector.render renders to document 1`] = `"<p>my foo: this is my foo, mybar: this is my bar</p>"`;

exports[`Connector.render works correctly 1`] = `
VNode {
  "attributes": Object {
    "baz": "qux",
    "foo": "bar",
  },
  "children": Array [],
  "key": undefined,
  "nodeName": [Function],
}
`;
